include/oneapi/dpl/experimental/kt/single_pass_scan.h

```diff
@@ -0,0 +1,413 @@
+          typename _TileVals>
+struct __lookback_kernel_func
+{
+    using _FlagStorageType = typename _FlagType::_FlagType;
+    static constexpr std::uint32_t __elems_in_tile = __workgroup_size * __data_per_workitem;
+
+    _InRng __in_rng;
+    _OutRng __out_rng;
+    _BinaryOp __binary_op;
+    ::std::size_t __n;
+    _StatusFlags __status_flags;
+    ::std::size_t __status_flags_size;
+    _StatusValues __status_vals_full;
+    _StatusValues __status_vals_partial;
+    ::std::size_t __current_num_items;
+    _TileVals __tile_vals;
+
+    [[sycl::reqd_sub_group_size(SUBGROUP_SIZE)]] void
+    operator()(const sycl::nd_item<1>& __item) const
+    {
+        auto __group = __item.get_group();
+        auto __subgroup = __item.get_sub_group();
+        auto __local_id = __item.get_local_id(0);
+
+        ::std::uint32_t __tile_id = 0;
+
+        // Obtain unique ID for this work-group that will be used in decoupled lookback
+        if (__group.leader())
+        {
+            sycl::atomic_ref<_FlagStorageType, sycl::memory_order::relaxed, sycl::memory_scope::device,
+                             sycl::access::address_space::global_space>
+                __idx_atomic(__status_flags[__status_flags_size - 1]);
+            __tile_id = __idx_atomic.fetch_add(1);
+        }
+
+        __tile_id = sycl::group_broadcast(__group, __tile_id, 0);
+
+        ::std::size_t __current_offset = static_cast<::std::size_t>(__tile_id) * __elems_in_tile;
+        auto __out_begin = __out_rng.begin() + __current_offset;
+
+        if (__current_offset >= __n)
+            return;
+
+        // Global load into local
+        auto __wg_current_offset = (__tile_id * __elems_in_tile);
+        auto __wg_next_offset = ((__tile_id + 1) * __elems_in_tile);
+        auto __wg_local_memory_size = __elems_in_tile;
+
+        if (__wg_next_offset > __n)
+            __wg_local_memory_size = __n - __wg_current_offset;
+
+        if (__wg_next_offset <= __n)
+        {
+            _ONEDPL_PRAGMA_UNROLL
+            for (std::uint32_t __i = 0; __i < __data_per_workitem; ++__i)
+            {
+                __tile_vals[__local_id + __workgroup_size * __i] =
+                    __in_rng[__wg_current_offset + __local_id + __workgroup_size * __i];
+            }
+        }
+        else
+        {
+            _ONEDPL_PRAGMA_UNROLL
+            for (std::uint32_t __i = 0; __i < __data_per_workitem; ++__i)
+            {
+                if (__wg_current_offset + __local_id + __workgroup_size * __i < __n)
+                {
+                    __tile_vals[__local_id + __workgroup_size * __i] =
+                        __in_rng[__wg_current_offset + __local_id + __workgroup_size * __i];
+                }
+            }
+        }
+
+        auto __tile_vals_ptr = __dpl_sycl::__get_accessor_ptr(__tile_vals);
+        _Type __local_reduction =
+            sycl::joint_reduce(__group, __tile_vals_ptr, __tile_vals_ptr + __wg_local_memory_size, __binary_op);
```

Comment:
We do have custom implementations of our own work-group reduce written by @julianmi. It was written specifically for the device-wide reduce algorithm and I haven't tried to incorporate it into this algorithm, although that would be a good future direction.

My thought process is that the work-group reduce is a rather small portion of the algorithm in terms of runtime conceptually, yet I do not have any data breaking down the runtime percentages so this could be an incorrect statement. Providing a breakdown of the percentages here would be helpful and is something that I can explore.

As far as a custom work-group scan, we have one in the main scan algorithm, but it exists as an alternative for the `sycl::joint_` scans when they cannot be used due to constraints of identity values / value types. From my understanding, it was written for the purpose of improving compatibility when the `sycl::joint_` scan cannot be used rather than for performance.

Reply:
Here are my thoughts on reduce: We have a single work-group implementation that would be closest to `sycl::joint_reduce`. I don't expect performance benefits of our generalized implementation over the highly optimized SYCL function though. The typical SLM required is the number of work-items times the data type size. Reusing this temporary memory could be beneficial.

Reply:
I'm unaware of documentation regarding the SLM requirements of our internal implementation. My statement concerning the SLM requirements was solely for our internal implementation, which uses the number of work-items times the data type size SLM space. 

I don't see a comment regarding SLM in the SYCL spec for `sycl::joint reduce`. It looks like the Intel compiler doesn't use SLM but instead a local reduction followed by a `reduce_over_group`, which typically uses shuffle and similar instructions to exchange data between work-items. See: https://github.com/intel/llvm/blob/sycl/sycl/include/sycl/group_algorithm.hpp#L349-L354.

I'm not sure if the SLM can be reused. I suspect that this depends on how the runtime optimizes SLM allocation and deallocation.